### PR TITLE
#3165: support for array to collection with adder preferred

### LINKED ISSUE #3129: CollectionMappingStrategy.ADDER_PREFERRED is not supported for Iterables
### Expected behavior

Having a source of type:
```java
public class Source {
    private java.util.List<String> edges;
    private java.util.List<String> points;

    public Source(List<String> edges, List<String> points) {
        this.edges = edges;
        this.points = points;
    }

    public List<String> getEdges() {
        return edges;
    }

    public List<String> getPoints() {
        return points;
    }
}
```
and a target of type:
```java
@lombok.Builder
@lombok.Value
public class Target{
    final @lombok.Singular("addEdge") com.google.common.collect.ImmutableList<String> edges;
    final @lombok.Singular("addPoint") java.util.List<String> points;
}
```
and a mapper interface:
```java
import org.mapstruct.CollectionMappingStrategy;
import org.mapstruct.Mapper;

@Mapper(collectionMappingStrategy = CollectionMappingStrategy.ADDER_PREFERRED)
public interface PathMapper {
    Target map(Source source);
}
```

, the mapper implementation uses `target.addEdge` to copy edges because of `CollectionMappingStrategy.ADDER_PREFERRED` is used.

### Actual behavior

`target.addEdge` is not used by the generated mapper implementation. Note that but `target.addPoint` is used as expected.

Mapper implementation code:
```java
public class MapperImpl implements Mapper {
    public MapperImpl() {
    }

    public Target map(Source source) {
        if (source == null) {
            return null;
        } else {
            Target.TargetBuilder target = Target.builder();
            target.edges(source.getEdges());
            if (source.getPoints() != null) {
                Iterator var3 = source.getPoints().iterator();

                while(var3.hasNext()) {
                    String point = (String)var3.next();
                    target.addPoint(point);
                }
            }

            return target.build();
        }
    }
}
```

### Steps to reproduce the problem

Use the `Source`, `Target` and `Mapper` classes to reproduce the issue. Lombok 1.18.24 was used with a default `lombok.singular.useGuava=false` setting. Note that  `lombok.singular.useGuava=true` would cause that both target properties (`edges`, `points`) will use [guava immutables](https://projectlombok.org/features/Builder) and the mapper implementation will skip usage of `target.addPoint` contrary to the expectation.

The issue is also reproducible without lombok, the actual Target class was transpiled by lombok to: 
```java
// unrelated methods to this issue (equals,hashcode ...)  were removed
import com.google.common.collect.ImmutableList;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public final class Target {
    private final ImmutableList<String> edges;
    private final List<String> points;

    Target(ImmutableList<String> edges, List<String> points) {
        this.edges = edges;
        this.points = points;
    }
    public static TargetBuilder builder() {
        return new TargetBuilder();
    }
    public ImmutableList<String> getEdges() {
        return this.edges;
    }
    public List<String> getPoints() {
        return this.points;
    }
    public static class TargetBuilder {
        private ImmutableList.Builder<String> edges;
        private ArrayList<String> points;

        TargetBuilder() {
        }

        public TargetBuilder addEdge(String addEdge) {
            if (this.edges == null) {
                this.edges = ImmutableList.builder();
            }

            this.edges.add(addEdge);
            return this;
        }

        public TargetBuilder edges(Iterable<? extends String> edges) {
            if (edges == null) {
                throw new NullPointerException("edges cannot be null");
            } else {
                if (this.edges == null) {
                    this.edges = ImmutableList.builder();
                }

                this.edges.addAll(edges);
                return this;
            }
        }

        public TargetBuilder addPoint(String addPoint) {
            if (this.points == null) {
                this.points = new ArrayList();
            }

            this.points.add(addPoint);
            return this;
        }

        public TargetBuilder points(Collection<? extends String> points) {
            if (points == null) {
                throw new NullPointerException("points cannot be null");
            } else {
                if (this.points == null) {
                    this.points = new ArrayList();
                }

                this.points.addAll(points);
                return this;
            }
        }

        public Target build() {
            ImmutableList<String> edges = this.edges == null ? ImmutableList.of() : this.edges.build();
            List points;
            switch (this.points == null ? 0 : this.points.size()) {
                case 0:
                    points = Collections.emptyList();
                    break;
                case 1:
                    points = Collections.singletonList((String)this.points.get(0));
                    break;
                default:
                    points = Collections.unmodifiableList(new ArrayList(this.points));
            }

            return new Target(edges, points);
        }
    }
}
```

### MapStruct Version

1.5.3.Final

### LINKED ISSUE #3165: Array to List mapping with adder fails
### Expected behavior

I would expect that given 
```java
public class User {
    private String id;
    private String email;
    private Permission[] permissions;
    // more code omitted for brevity
}
```
I'd be able to get a mapper to map the Java POJO to
```
message UserDTO {
    string id = 1;
    string email = 2;
    repeated PermissionDTO permissions = 3;
// more code omitted for brevity
}
```
Which  would map the `permission` array to the proto `permissionsList` through the adder pattern.

### Actual behavior

The auto-code generation fails because the generated code tries to directly add the whole array instead of iterating over it, map each value on its own and add it to the list.

### Steps to reproduce the problem

I modified the public Protobuff example so it fails
https://github.com/eroznik/mapstruct-examples/commit/f4c3a4eae7a678b8ba83820bdee4532f6d02b622

The pushed code is using the latest snapshot, mostly because I was using the latest snapshot version while testing it out locally. Keep in mind that the linked commit won't build.

### MapStruct Version

All versions that I tested.

### PR DESCRIPTION
**Pull request by eroznik**
This PR offers a potential fix for the problem that I described in #3165 .

Fixes #3165 
Fixes #3129

### CODE REVIEW

**Comment by filiphr** [comment]
This fix looks OK. However, there are tests missing. If you would like for us to accept this PR please provide tests. 

By the way the support is not really linked to protobuf. It is rather linked to not being able to map an array to a collection / iterable using an adder. It would be good to add tests like that (not with protobuf).

**Comment by eroznik** [comment]
@filiphr i've just added a test case for the issue, lmk if thats what you expected.. (:

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/AdderWrapper.java [hunk 1]
```diff
@@ -39,7 +39,12 @@ public AdderWrapper( Assignment rhs,
         // localVar is iteratorVariable
         String desiredName = Nouns.singularize( adderIteratorName );
         rhs.setSourceLoopVarName( rhs.createUniqueVarName( desiredName ) );
-        adderType = first( getSourceType().determineTypeArguments( Collection.class ) );
+        if ( getSourceType().isCollectionType() ) {
+            adderType = first( getSourceType().determineTypeArguments( Collection.class ) );
+        }
+        else { // array
+            adderType =  getSourceType().getComponentType();
+        }
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/assignment/AdderWrapper.java:47]: This won't work if the source type is iterable i.e. `Iterable<String>`. Can you please add a test case for this as well?
```suggestion
        if ( getSourceType().isArrayType() ) {
            adderType  =  getSourceType().getComponentType();
        }
        else { // iterable
            adderType = first( getSourceType().determineTypeArguments( Collection.class ) );
            adderType =  getSourceType().getComponentType();
        }
```

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/SourceRHS.java [hunk 2]
```diff
@@ -149,6 +149,9 @@ public Type getSourceTypeForMatching() {
             else if ( sourceType.isStreamType() ) {
                 return first( sourceType.determineTypeArguments( Stream.class ) );
             }
+            else if ( sourceType.isArrayType() ) {
+                return sourceType.getComponentType();
+            }
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/SourceRHS.java:154]: I think that we might need to have another `else if ( sourceType.isIterableType() )` here as well.

**Review by filiphr — COMMENTED** [review]
The test you added is good. Would be nice if you can also add a test where the source is `Iterable<String>`. I think that it won't work without the changes I am mentioning in the comments

**Comment by filiphr** [comment]
I think that with the addition of the other tests we are going to show that this is also going to fix #3129 and actually make PR #3130 obsolete

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/SourceRHS.java (hunk 2 above)
Comment (eroznik) [reply processor/src/main/java/org/mapstruct/ap/internal/model/common/SourceRHS.java:154]: you are correct :+1:

File: processor/src/main/java/org/mapstruct/ap/internal/model/assignment/AdderWrapper.java (hunk 1 above)
Comment (eroznik) [reply processor/src/main/java/org/mapstruct/ap/internal/model/assignment/AdderWrapper.java]: added

**Comment by eroznik** [comment]
Added test case for iterable mapping + the required code changes so the test passes

**Comment by filiphr** [comment]
Thanks @eroznik

**Pull request merged**